python_benchmarks/core.py

```diff
@@ -0,0 +1,202 @@
+        """
+        try:
+            self.prof.stop()
+            prof_averages = self.prof.key_averages()
+            elapsed_cuda_time = self._get_kernel_time(prof_averages)
+            self._increment_global_time(elapsed_cuda_time)
+            self.prof.start()
+        except:
+            self.prof.start()
+        return self.current_time
+
+    def _get_kernel_time(self, prof_averages) -> float:
+        """
+        Returns:
+            time_value: Elapsed CUDA time in seconds.
+        """
+        elapsed_cuda_time = sum([event.self_cuda_time_total
+            for event in prof_averages 
+            if event.device_type == DeviceType.CUDA])/ 1e6
+        return elapsed_cuda_time
+
+    def _increment_global_time(self, elapsed_time: float) -> None:
+        self.current_time += elapsed_time
+
+    def cleanup(self) -> None:
+        """
+        Stops a running torchprofiler instance if found.
+        """
+        try:
+            self.prof.stop()
+        except:
+            pass
+
+    def compute_metrics(
+        self, inputs: Union[torch.Tensor, List], outputs: Union[torch.Tensor, List]
+    ) -> None:
+        """
+        Utility function to compute metrics for the target function.
+        Current metrics:
+            IOBytes: Total bytes in inputs + outputs
+            BytesPerSecond: IOBytes * total_rounds / total_time
+            Bandwdith (GBps): BytesPerSecond / (1024**3)
+            % Peak Bandwidth (SOL): 100 * Bandwidth /PEAK_BANDWIDTH
+        """
+        if isinstance(inputs, torch.Tensor):
+            inputs = [inputs]
+        if isinstance(outputs, torch.Tensor):
+            outputs = [outputs]
+
+        iobytes = 0
+        for inp in inputs:
+            if isinstance(inp, torch.Tensor):
+                iobytes += inp.element_size() * inp.numel()
+        for out in outputs:
+            if isinstance(out, torch.Tensor):
+                iobytes += out.element_size() * out.numel()
+
+        self.benchmark.extra_info["IOBytes"] = iobytes
+        bytes_per_second = (
+            iobytes * self.benchmark.stats["rounds"]
+        ) / self.benchmark.stats["total"]
+        self.benchmark.extra_info["BytesPerSecond"] = bytes_per_second
+        self.benchmark.extra_info["Bandwidth (GBps)"] = bytes_per_second / 1024**3
+        self.benchmark.extra_info["% Peak Bandwidth (SOL)"] = (
+            100 * (bytes_per_second / 1024**3) / PEAK_BANDWIDTH
+        )
+
+
+def run_benchmark(
+    benchmark, benchmark_fn: Callable, inputs: Union[torch.Tensor, List]
+) -> Union[torch.Tensor, List]:
+    """
+    Benchmarks the target function using torchprofiler and stores metrics as extra information.
+
+    Arguments:
+        benchmark: pytest-benchmark fixture
+        benchmark_fn: Target function
+        inputs: Inputs to the target function
+
+    Returns:
+        outputs: Output of the target function
+    """
+    nvf_bench = NVFBenchmark(benchmark)
+    clear_l2_cache()
```

Comment:
Can we move `clear_l2_cache` into the constructor? This way, we don't need to expose it as a public method.

Reply:
My reason for keeping it as a separate call was to make sure that we clear the cache right before running the benchmarks in case we have additional setup steps (storing any further benchmark information etc.) before running the benchmark.

Reply:
That's fair. We can revisit the API in a separate PR. Looking at the use cases so far, it makes more sense to wrap the benchmark framework as a context manager so users can easily scope the region to benchmark.